#1830 implement {DataFrame, Series}.empty

Fernadoo opened the pull request:
<!--
Thank you for your contribution!

Please review https://github.com/mars-project/mars/blob/master/CONTRIBUTING.rst before opening a pull request.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

I implemented .empty property for {DataFrame, Series}.

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->

Issue #1830 should be resolved

hekaisheng commented:
Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?

Fernadoo commented:
> Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?

Sure, I am gonna do it later today :D

hekaisheng commented:
> > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> 
> Sure, I am gonna do it later today :D

BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.

Fernadoo commented:
> > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > 
> > 
> > Sure, I am gonna do it later today :D
> 
> BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.

Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?

qinxuye commented:
> > > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > > 
> > > 
> > > Sure, I am gonna do it later today :D
> > 
> > 
> > BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.
> 
> Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?

@hekaisheng means commit https://github.com/mars-project/mars/commit/37455c1007ca1d255de9ca789de5da70cbf1864e, just rebase the master branch please.

Fernadoo commented:
> > > > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > > > 
> > > > 
> > > > Sure, I am gonna do it later today :D
> > > 
> > > 
> > > BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.
> > 
> > 
> > Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?
> 
> @hekaisheng means commit [37455c1](https://github.com/mars-project/mars/commit/37455c1007ca1d255de9ca789de5da70cbf1864e), just rebase the master branch please.

Oh, I see! Sure then.

Fernadoo commented:
> > > > > > Thanks for your contribution! Could you add test cases in `dataframe.base.tests` ?
> > > > > 
> > > > > 
> > > > > Sure, I am gonna do it later today :D
> > > > 
> > > > 
> > > > BTW, we merged a large pull request, please rebase it first and adapt to the new style of tests.
> > > 
> > > 
> > > Excuse me, I got a bit confused. The commit history shows that you have merged the latest PR into v0.6. Nevertheless, according to the contributing guideline, I was developing upon master and submitted my PR against master as well. So do you mean that I am indeed supposed to rebase v0.6 to my private branch and also develop upon v0.6?
> > 
> > 
> > @hekaisheng means commit [37455c1](https://github.com/mars-project/mars/commit/37455c1007ca1d255de9ca789de5da70cbf1864e), just rebase the master branch please.
> 
> Oh, I see! Sure then.

Since `pytest mars` did not work well under my forked repo, so I just downloaded [`mars-project:master`](https://github.com/mars-project/mars) without any modification to evaluate all test cases. It is supposed to pass all cases. However, it turns out that the command `pytest mars` will always be stuck in `mars/dataframe/arithmetic/tests/test_arithmetic_execution.py`, attached as the figure below (sometimes even outputs an 'F' and stops moving on):
<img width="719" alt="pytest_err" src="https://user-images.githubusercontent.com/42331572/121197315-cd07d880-c8a3-11eb-8c7a-20bb40165b35.png">
I am wondering if this situation has happened before?

hekaisheng commented:
Could you try to run `python setup.py build_ext -i` and run tests again?  It will takes a long time to check all cases, just make sure the cases you added are pass.

qinxuye commented:
@hekaisheng is right, you just need to pay attention to the test you added, the entire unittests are too huge to cover.

Fernadoo commented:
Hi @hekaisheng and @qinxuye, I have added some test cases in `mars/dataframe/base/tests/test_base.py`. The pytest results show that those newly added cases are passed. Is there anything left that I can help with? 
(BTW, I have already rebased upon your latest commit [`2805241`](https://github.com/mars-project/mars/commit/2805241ac55b50c4f6319baa41113fbf8c723832), no worries)

hekaisheng commented:
Could you test with this case:
``` Python
a = md.DataFrame(np.random.rand(10, 2))
assert a[a > 0].empty
```

In some cases, Mars tensor or DataFrame has unknown shape as Mars is a lazy evaluation framework, its shape can't be decided before execution.

hekaisheng commented on mars/dataframe/core.py:
```diff
@@ -1021,6 +1021,18 @@ def index(self):
         return [self.index]
 
+    """
+    Property .empty STARTS
```
Usually, we use git blame to show the the revision and author of code, comments are not necessary.

Fernadoo replied:
Sure then :D

Fernadoo commented:
> Could you test with this case:
> 
> ```python
> a = md.DataFrame(np.random.rand(10, 2))
> assert a[a > 0].empty
> ```
> 
> In some cases, Mars tensor or DataFrame has unknown shape as Mars is a lazy evaluation framework, its shape can't be decided before execution.

Your conjecture is right, `a[a > 0].empty` does not return `True`. Let me fix this.

hekaisheng commented:
We implement `mt.shape` (mars/tensor.shape) to get shape of those tensors has unknown shape, if mars.dataframe.shape is implemented, `df.empty` can return a Mars tensor and get value after triggering execution.

hekaisheng commented:
For this PR, you can just raise an error if DataFrame's shape has `np.nan`, and tell users to trigger execution first, its shape will be updated after call `df.execute()`.

Fernadoo commented:
Hi @hekaisheng, I have added `raise` module and deleted unnecessary comments. Plz have a look:D

hekaisheng commented on mars/dataframe/base/tests/test_base.py:
```diff
@@ -888,3 +888,21 @@ def a():
+    assert from_pandas_series(pd.Series({'a':[1]})).empty == pd.Series({'a':[1]}).empty
+
+    # Maybe fail due to lazy evaluation
+    # assert from_pandas_df(pd.DataFrame(np.random.rand(10, 2))).empty == True
```
Use `pytest.raises` to test unknown shape.

Fernadoo replied:
Resolved!

hekaisheng commented on mars/dataframe/core.py:
```diff
@@ -1499,6 +1506,13 @@ def columns(self):
+        shape = getattr(self, '_shape')
+        if np.nan in shape:
+            raise RuntimeError("The shape is unknown until you trigger execution")
+        return shape[0] == 0
```
if length of columns is 0, `df.empty` also returns True

Fernadoo replied:
Resolved

Fernadoo commented:
Please have a look if it's ok now :D, thanks for your patience!

hekaisheng commented on mars/dataframe/base/tests/test_base.py:
```diff
@@ -888,3 +888,24 @@ def a():
+    # Maybe fail due to lazy evaluation
+    with pytest.raises(RuntimeError):
+        a = from_pandas_df(pd.DataFrame(np.random.rand(10, 2)))
+        assert a[a>0].empty == True
```
Add a new line at end of file

Fernadoo replied:
Have done it but don't know why we need it

Fernadoo replied:
It seems that I also need to modify a[a>0] to a[a > 0], because of [ubuntu-latest, 3.9-cython] tests?

hekaisheng replied:
We use Flake8 for code linting, it breaks the rule `No newline at end of file (W292)`.

Fernadoo replied:
Then let me fix this and `reset` to make the log a bit cleaner

qinxuye commented on mars/dataframe/core.py:
```diff
@@ -1021,6 +1021,13 @@ def index(self):
+    def empty(self):
+        shape = getattr(self, '_shape')
+        if np.nan in shape:
+            raise RuntimeError("The shape is unknown until you trigger execution")
```
`ValueError` can be raised which can refer to https://github.com/mars-project/mars/blob/6535fe1b198e2ef7b209058d442a3b3f375cfba5/mars/core/entity/executable.py#L90-L91 .

hekaisheng commented on mars/dataframe/core.py:
```diff
@@ -1499,6 +1506,13 @@ def columns(self):
+    @property
+    def empty(self):
+        shape = getattr(self, '_shape')
+        if np.nan in shape:
```
It's better to use `np.isnan` to check.

qinxuye reviewed (commented):
Code in `actors` directory need to be removed.

hekaisheng reviewed (changes requested):
Could you check your changes to make sure other files are not included?

Fernadoo commented:
@hekaisheng @qinxuye, sure, also notice that, don't know why but I shall fix these

Fernadoo commented:
New implementation redirected to [#2155](https://github.com/mars-project/mars/pull/2155)
